sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -14,197 +16,180 @@ from ...cluster._hierarchical_fast cimport UnionFind
 from ...utils._typedefs cimport ITYPE_t, DTYPE_t
 from ...utils._typedefs import ITYPE, DTYPE
 
-cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_from_distance_matrix(
-    cnp.ndarray[cnp.double_t, ndim=2] distance_matrix
+# Numpy structured dtype representing a single ordered edge in Prim's algorithm
+MST_edge_dtype = np.dtype([
+    ("current_node", np.intp),
+    ("next_node", np.intp),
+    ("distance", np.float64),
+])
+
+# Packed shouldn't make a difference since they're all 8-byte quantities,
+# but it's included just to be safe.
+ctypedef packed struct MST_edge_t:
+    cnp.intp_t current_node
+    cnp.intp_t next_node
+    cnp.float64_t distance
+
+cpdef cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst_from_mutual_reachability(
+    cnp.ndarray[cnp.float64_t, ndim=2] mutual_reachability
 ):
```

Comment:
We should only use `cnp.ndarray` when memoryviews can't be used.

Is the usage of `MST_edge_t` preventing us to use (`const`-qualified) memoryview here and in other functions and methods?

Generally, most `cnp.ndarray` seem replaceable by memoryviews here.

Reply:
The current algorithm uses binary masks for indexing into `cnp.ndarray`s, simplifying the innermost scope of the algorithm. Hence I think it would be preferable to maintain the use of `cnp.ndarray` instead. In particular, it is necessary for:
https://github.com/scikit-learn/scikit-learn/blob/39e7d7e59dfdcf52488bd0d05dc7c43e67becef2/sklearn/cluster/_hdbscan/_linkage.pyx#L48-L52

Reply:
I could attempt to change the algorithm to avoid using this, but it is a pretty eloquent solution. Ultimately I don't know _how costly_ the python interactions here are, and whether they outweigh the usefulness of the algorithm as it is now.

Reply:
There are some overhead of just typing and referencing variable with `cnp.ndarray` instead of using memoryviews (because `cnp.ndarray` are `PyObjects` and memoryviews aren't if I remember correctly).

We can leave this for now. If this shows to be a hotspot, we can have a PR to rewrite this part. What do you think?

In this case, is it possible to add a comment to indicate that numpy arrays are used for binary masks and convenient indexing?

Reply:
Sounds good, I will add in such a comment.